opnieuw/retries.py

```diff
@@ -220,87 +224,77 @@ def foo() -> None:
+                    MonotonicClock(),
+                    max_calls_total=max_calls_total,
+                    retry_window_after_first_call_in_seconds=retry_window_after_first_call_in_seconds,
+                )
+
+                last_exception = None
+                while True:
+                    try:
+                        # Mypy currently does not propagate type-narrowing info from outside this asyc def
+                        # to its body.
+                        # Therefore, the following cast is necessary.
+                        # You can find a longer writeup on StackOverflow: https://stackoverflow.com/a/75439065/1067339
+                        # The linked-to issue (https://github.com/python/mypy/issues/2608) is nowadays closed;
+                        # MyPy accepts the following in non-strict mode but the cast is still necessary in strict mode.
+                        #
+                        # This cast is sound because of the `inspect.iscoroutinefunction(f)` above.
+                        return await cast(Awaitable[R], f(*args, **kwargs))
+                    except Exception as e:
+                        if last_exception is not None:
+                            e.__cause__ = last_exception
+
+                        last_exception = e
+                        if not isinstance(e, retry_on_exceptions):
+                            raise
+
+                        if (sleep_seconds := backoff_calculator.get_backoff()) is None:
+                            raise
+
+                        await asyncio.sleep(sleep_seconds)
+            return functools.wraps(f)(async_wrapper)
+        else:
+            def sync_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+                backoff_calculator = _get_backoff_calculator_class(namespace)(
+                    MonotonicClock(),
+                    max_calls_total=max_calls_total,
+                    retry_window_after_first_call_in_seconds=retry_window_after_first_call_in_seconds,
+                )
+
+                last_exception = None
+                while True:
+                    try:
+                        return f(*args, **kwargs)
+                    except Exception as e:
+                        if last_exception is not None:
+                            e.__cause__ = last_exception
+
+                        last_exception = e
+                        if not isinstance(e, retry_on_exceptions):
+                            raise
+
+                        if (sleep_seconds := backoff_calculator.get_backoff()) is None:
+                            raise
+
+                        time.sleep(sleep_seconds)
+            return functools.wraps(f)(sync_wrapper)
     return decorator
 
-
-retry_async.__doc__ = retry.__doc__
+# We expose `retry_async` for backwards-compatibility.
+# However, nowadays the main `retry` decorator is able
+# to accept both sync and async functions directly.
+retry_async = retry
```

Comment:
It would be nice to mark `retry_async` with `warnings.deprecated` to indicate we can get rid of this one in the future. I don't think we can use `typing.deprecated` yet without taking `typing_extensions` as a dependency.

Reply:
Would it be acceptable to add `typing_extensions` as extra dependency? Especially seeing as it is a backport of functionality that is built into newer Python versions (which people are much more likely to use than say 3.8) anyway?